src/Cms/Environment.php

```diff
@@ -39,62 +39,23 @@ public function __construct(string $root, $allowed = null)
     {
         $this->root = $root;
 
-        // This is the default. The current URL should be detected via the server name.
-        // Empty hosts are allowed. Should lead to no breaking changes with previous setups.
-        if ($allowed === null) {
-            Server::hosts([]);
-            $this->uri = Uri::index();
-
-        // The current URL should be detected via possibly insecure HOST headers
-        } elseif ($allowed === true) {
-            Server::hosts(['*']);
-            $this->uri = Uri::index();
-            $this->blockEmptyHost();
-
-        // the current URL should be detected via the server name. no empty hosts allowed
-        } elseif ($allowed === false) {
-            Server::hosts([]);
-            $this->uri = Uri::index();
-            $this->blockEmptyHost();
-
-        // the current URL is predefined with a single string
-        // and not detected automatically.
-        // If the url option is relative (i.e. '/' or '/some/subfolder')
-        // the host will be empty and that's totally fine.
-        // No need to block an empty host here
-        } elseif (is_string($allowed) === true) {
-            // create the URI object directly from the given option
-            // without any form of detection from the server
-            $this->uri = new Uri($allowed);
-
-            // only create an allow list from absolute URLs
-            // otherwise the default secure host detection
-            // behavior will be used
-            if (empty($host = $this->uri->host()) === false) {
-                Server::hosts([$host]);
-            }
-
-            // the current URL should be auto detected from a host allowlist
-        } elseif (is_array($allowed) === true) {
-            foreach ($allowed as $url) {
-                $host    = (new Uri($url))->host();
-                $hosts[] = $host;
-            }
-
-            // register all allowed hosts
-            Server::hosts($hosts);
-
-            // get the index URL, including the subfolder if it exists
-            $this->uri = Uri::index();
-            $this->blockEmptyHost();
-
-            // validate against the list of allowed base URLs
-            if (in_array((string)$this->uri, $allowed) === false) {
-                throw new InvalidArgumentException('The subfolder is not in the allowed base URL list');
-            }
-        } else {
-            throw new InvalidArgumentException('Invalid allow list setup for base URLs');
+        if (is_string($allowed) === true) {
+            return $this->setupFromString($allowed);
```

Comment:
One detail: The constructor should not return values, might trigger deprecation warnings in the future.

Either the methods should return `void` or the constructor should use a pattern like this:

```php
$this->setupFromString($allowed);
return;
```